src/v/pandaproxy/schema_registry/store.h

```diff
@@ -326,10 +392,17 @@ class store {
     }
 
     bool upsert_subject(
-      subject sub, schema_version version, schema_id id, is_deleted deleted) {
+      seq_marker marker,
+      subject sub,
+      schema_version version,
+      schema_id id,
+      is_deleted deleted) {
```

Comment:
This function is quite complex now. I wonder if this is equivalent:
```cpp
    bool upsert_subject(
      seq_marker marker,
      subject sub,
      schema_version version,
      schema_id id,
      is_deleted deleted) {
        auto& subject_entry = _subjects[std::move(sub)];
        auto& versions = subject_entry.versions;

        subject_entry.written_at.push_back(marker);

        const auto v_it = std::lower_bound(
          versions.begin(),
          versions.end(),
          version,
          [](const subject_version_id& lhs, schema_version rhs) {
              return lhs.version < rhs;
          });
        const bool found = v_it != versions.end() && v_it->version == version;
        if (found) {
            *v_it = subject_version_id(version, id, deleted);
        } else {
            versions.insert(v_it, subject_version_id(version, id, deleted));
        }

        // The subject is deleted iff all versions are deleted
        subject_entry.deleted = is_deleted(
          std::all_of(versions.begin(), versions.end(), [](const auto& v) {
              return v.deleted;
          }));

        return !found;
    }
```
It drops quite a few conditionals on updating `subject_entry.deleted`

Reply:
I think ti might miss the case where the subject is just deleted, so perhaps:
```cpp
subject_entry.deleted = is_deleted(subject_entry.deleted ||
          std::all_of(versions.begin(), versions.end(), [](const auto& v) {
              return v.deleted;
          }));
```

Reply:
I've gone with mostly what you suggest, plus a twiddle to the subject deletion logic to only update the subjects `deleted` member if the `all_of` result agrees with the direction we're going (`deleted`), to avoid possibility of incorrectly un-deleting a subject.